### PR TITLE
Dockerize app configuration and dependent scripts

### PR DESCRIPTION
**Pull request by brew**
This PR further containerises the os-admin app, and tidies the Dockerfile statements and start script.

- The main change is removing the static `config.js` file, and using the node server to return a json response to `/config` endpoint. This json response is dynamically created from available settings in the environment and, if present, `.env` file.

- The configuration is loaded early on in the application lifecycle, so settings are available to the Angular app, and other dependent libraries.

- Dependent libraries, `authClient.services` and `snippets` were previously hardcoded into the app html template. These have been removed, and instead are dynamically loaded by the app bootstrapping process after config has been successfully loaded.

- This PR uses [lookenv](https://github.com/RodrigoEspinosa/lookenv) to define which env settings the app uses, to ensure all required settings are present, and to load `.env` settings if present.

Part of openspending/openspending#1340.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,9 +1,17 @@
...
-RUN git clone http://github.com/openspending/os-admin.git app
-RUN cd app && npm install
-RUN cd app && npm run build
+RUN apk add --update git
+
+# use changes to package.json to force Docker not to use the cache
+# when we change our application's nodejs dependencies:
+ADD package.json /tmp/package.json
+RUN cd /tmp && npm install
+RUN mkdir -p /opt/app && cp -a /tmp/node_modules /opt/app/
```
Comment (vitorbaptista) [inline Dockerfile]: Why are you using `/tmp`? Why not simply

```
WORKDIR /app
ADD package.json .
RUN npm install

ADD . .
```

(I used `/app` because I feel that's what's used in the other Dockerfiles, but could be `/opt/app` as well)

File: README.md [hunk 2]
```diff
@@ -16,6 +16,25 @@ OpenSpending user administration page
...
+OS_SNIPPETS_GA= # Google Analytics code
+OS_SNIPPETS_RAVEN= # Sentry public DSN url
+
+# Each service will use OS_BASE_URL unless overridden by these:
+OS_CONDUCTOR_URL=
+OS_SEARCH_URL=
+OS_VIEWER_URL=
+OS_EXPLORER_URL=
+OS_PACKAGER_URL=
+```
```
Comment (vitorbaptista) [inline README.md:39]: Instead of adding it in the README (or in addition to), I find it useful to have a `.env.example` file with some sane defaults. Just keep in mind that if you have `FOOBAR= # Add foobar`, the contents of `FOOBAR` will be `" # Add foobar"`. Comments only work on their own separate line.

File: app/scripts/index.js [hunk 3]
```diff
@@ -18,6 +18,49 @@ var _ = require('lodash');
 
   require('os-bootstrap/dist/js/os-bootstrap');
 
+  // Make config global available to app, and load scripts that require
+  // globalConfig.
+  $.get('./config')
+  .then(function(res) {
```
Comment (vitorbaptista) [inline app/scripts/index.js]: It might be easier to have the path `config.json`, and then you'd run:

```javascript
$.get('config.json')  // the `./` is unecessary
  .then(function(config) {
    // The config will already have the logic of extracting env variables to their specific config keys

    globals.globalConfig = Object.assign({}, globals.globalConfig, config);

    return globals.globalConfig;
  });
```

The logic of extracting the environment variables to config keys would be implemented in the `config.json` route, similar to how it was done with the `config.js`.

File: server.js [hunk 4]
```diff
@@ -1,11 +1,27 @@
...
 var express = require('express');
 var app = express();
+var env = require('dotenv').load();
+var envConfig = require('./lookenv.config');
+
 
 app.use(express.static('.'));
 
-app.listen(8000, function () {
+app.get('/config', function(request, response) {
```
Comment (vitorbaptista) [inline server.js]: As it's a JSON, I think having the route as `/config.json` is more explicit.

Instead of simply being a dump generator of a JSON with the environment keys, why not generate the JSON with the actual config you want? Like:

```javascript
var config = {
    conductorUrl: envConfig.OS_CONDUCTOR_URL || envConfig.OS_BASE_URL,
    explorerUrl: envConfig.OS_EXPLORER_URL || envConfig.OS_BASE_URL,
    // ...
}
```

Basically, this route returns the config.json as we want it to be, so any place that wants a config can use it (including ourselves, if we want to debug things)

Comment (vitorbaptista) [reply server.js]: The main advantage I see in having this logic here, and not on the frontend, is that it's easier to test as needed. Not that I think we should test it all, but it might be useful for some parts of the logic, like the `OS_SNIPPETS_`.

**Review by vitorbaptista — COMMENTED** [review]
I added some comments to the PR. Your changes are the best way forward IMO, my comments are mostly small tweaks/suggestions.

Apart from them, I would suggest checking the indenting. There were a few places where the code was indented like:

```javascript
$.get('foo')
.then(function(res) {})
```

I prefer to keep the dots in the same level of indentation.

Great job on these changes :clap:

File: Dockerfile (hunk 1 above)
Comment (brew) [reply Dockerfile]: Yes, I agree the `/opt/app` juggling isn't adding anything to the cache optimisation.

Comment (pwalsh) [reply Dockerfile]: for reference, I use `/srv/app`, and that became a pattern on the CKAN images @amercader and I did a couple of years ago, and is also present in Adria's latest docker work eg https://github.com/okfn/docker-ckan/blob/master/ckan-base/Dockerfile#L4

File: server.js (hunk 4 above)
Comment (akariv) [reply server.js]: A different approach I've used a couple of times is to embed the configuration inside a script element in index.html:
```html
<script>
  window.globalConfig = {{ config || safe }};
</script>
```
And then let node render index.html with the configuration inside.

This allows us to remove the `/config` endpoint, and, more importantly, simplifies the frontend app which may rely on the configuration always being there instead of having to asynchronously fetch it  (and everything else initializing asynchronously after it etc.)

An example of this approach can be seen here: https://github.com/OpenBudget/budgetkey-app-generic-item/blob/master/index.js
Where node is prefetching some data and embedding a theme configuration in index.html before serving it.

Comment (vitorbaptista) [reply server.js]: Adding to that idea, we could instead have a `/config.js` route, similar to what we had with:

```js
window.globalConfig = {
  // ...
};
```

and then just add `<script src="config.js">`. This needs to be loaded before any other script, and we might have problems with busting the browser's cache, but then we can still use our current `app.use(express.static('.'));`.

**Comment by akariv** [comment]
Great suggestion @*vitorbaptista!*

File: server.js (hunk 4 above)
Comment (brew) [reply server.js]: I wanted to avoid having `window.globalConfig` defined outside of the responsibility of the Angular app. Defining it in the initialisation code centralises it with other global definitions. It makes the Angular app responsible for its globals, and doesn't presume it will just be there.

However, we can improve the current code by removing the chained `thens` after requesting the config, and just make the dependent script requests within the first `then`.

```javascript
  // Make config global available to app, then load scripts that require
  // globalConfig.
  $.get('config.json')
    .then(function(config) {
      globals.globalConfig = Object.assign({}, globals.globalConfig, config);

      // Load snippets (requires globalConfig)
      $.getScript('public/scripts/snippets.js')
        .fail(function(jqxhr, settings, exception) {
          throw(exception);
        });

      // Load externally hosted authClient.services lib (makes `authenticate`
      // and `authorize` services available to angular modules).
      var libUrl = globalConfig.conductorUrl + '/user/lib';
      $.getScript(libUrl)
        .fail(function(jqxhr, settings, exception) {
          console.log('Unable to load authClient.services from ' + libUrl);
        });

      return globals.globalConfig;
    });
```

Comment (vitorbaptista) [reply server.js]: Although having Angular take care of everything it needs is a good goal, I think these config variables can be treated differently. It's similar to environment variables: the code doesn't deal with how to load the env variables, they're just there, injected by someone else. The project is concerned with the code, but not the environment.

The advantage of dealing it this way is that it simplifies our code. For example, the current code doesn't deal with the case on when the `/config.json` path fails. If the config is instead injected via a `<script src="config.js">`, the `index.js` could contain code like:

```js
(function(globals) {
  if (globals.globalConfig === undefined) {
    throw "Missing globalConfig object in the global scope.";
  }
  // ...
})(window);
```

In other words, the `index.js` just ensures that the config was loaded, just to be a bit defensive.

Comment (brew) [reply server.js]: Having added these scripts via `$.getScript` in explorer now, I think @vitorbaptista is right. It's less error-prone and clearer to add `config.js` and `snippets.js` via normal script tags. While I don't like scripts relying on magic globals being present, it simplifies the initialisation process, as @akariv suggests. This also helps resolve a related issues, specific to os-explorer, where the `/config.json` path is being redirected to the legacy api in the ngnix location config.

Comment (brew) [reply server.js]: Thanks both for your comments and suggests here!
